activesupport/lib/active_support/hash_with_indifferent_access.rb

```diff
@@ -94,7 +94,11 @@ def self.[](*args)
     #
     # This value can be later fetched using either +:key+ or +'key'+.
     def []=(key, value)
-      regular_writer(convert_key(key), convert_value(value, for: :assignment))
+      if value.frozen?
+        regular_writer(convert_key(key), convert_value(value))
+      else
+        regular_writer(convert_key(key), convert_value(value, for: :assignment))
+      end
```

Comment:
I think this will create the duplicate of `value` whenever we are not passing `:for` key in the `options` hash, since `convert_value` is also called from [here](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/hash_with_indifferent_access.rb#L135).